File: starter_kits/Java/hlt/Direction.java
package hlt;

import java.util.ArrayList;

public enum Direction {
    NORTH('n'),
    EAST('e'),
    SOUTH('s'),
    WEST('w'),
    STILL('o');

    public final char charValue;

    public final static ArrayList<Direction> ALL_CARDINALS = new ArrayList<>();
    static {
        ALL_CARDINALS.add(NORTH);
        ALL_CARDINALS.add(SOUTH);
        ALL_CARDINALS.add(EAST);
        ALL_CARDINALS.add(WEST);
    }

    public Direction invertDirection() {
        switch (this) {
            case NORTH: return SOUTH;
            case EAST: return WEST;
            case SOUTH: return NORTH;
            case WEST: return EAST;
            case STILL: return STILL;
            default: throw new IllegalStateException("Unknown direction " + this);
        }
    }

    Direction(final char charValue) {
        this.charValue = charValue;
    }
}

File: starter_kits/Kotlin/hlt/Direction.kt
package hlt

import java.util.*

enum class Direction constructor(val charValue: Char) {
    NORTH('n'),
    EAST('e'),
    SOUTH('s'),
    WEST('w'),
    STILL('o');

    fun invertDirection(): Direction {
        when (this) {
            NORTH -> return SOUTH
            EAST -> return WEST
            SOUTH -> return NORTH
            WEST -> return EAST
            STILL -> return STILL
            else -> throw IllegalStateException("Unknown direction " + this)
        }
    }

    companion object {

        val ALL_CARDINALS: ArrayList<Direction> = ArrayList()

        init {
            ALL_CARDINALS.add(NORTH)
            ALL_CARDINALS.add(SOUTH)
            ALL_CARDINALS.add(EAST)
            ALL_CARDINALS.add(WEST)
        }
    }
}
